### DISCUSSION TITLE
Linux 编译流程

### DISCUSSION THREAD

**Discussion by xfangfang** [xfangfang/wiliwili#89] [Q&A]
不同Linux系统可能存在不同的依赖和配置参数，这里收集不同 Linux 下的编译流程

**Comment by xfangfang** [comment]
请在看这篇文档前先读过readme中的编译说明，下面的内容是补充

# 拉取代码
首先要注意如何拉取项目，因为项目下包含子模块，请一定要按照readme所说的方式拉取项目：

git clone --recursive https://github.com/xfangfang/wiliwili.git

# 编译依赖
支持编译 c++17 的编译工具，make（或ninja） 和 cmake 

x11 编译依赖或 wayland 编译依赖 

默认使用 x11，如果需要 wayland，则要添加 cmake 选项： -DGLFW_BUILD_WAYLAND=ON  -DGLFW_BUILD_X11=OFF

# 项目依赖
应用需要的外部依赖为 dbus, libmpv 和 tls 库（mbedtls, openssl, gnutls 之一即可）

可选依赖 webp。若构建时使用了webp，可以从设置中的开源声明里看到对应的内容。使用 webp 可以减少图片流量，提升加载速度。

如果您在构建过程中不能访问互联网，比如构建 flatpak 包，那么需要添加 cmake 选项 -DUSE_SYSTEM_CURL=ON 并且安装 libcurl 。 
如果构建的过程中可以访问互联网，则会在运行 cmake 指令时自动下载指定版本的 libcurl 源码用于稍后编译。

如果你使用了系统搜索路径之外的环境，比如：使用homebrew安装最新的libmpv，而不是使用系统包管理器安装的libmpv，可以在 cmake指令后添加额外的搜索路径：`cmake -B build -DPLATFORM_DESKTOP=ON -DCMAKE_PREFIX_PATH="$(brew --prefix)"`

**Reply by xfangfang** [reply]
@chenkaihuai 我刚刚查看大家 fork 的修改的时候注意到了您更新的内容，非常感谢。

特意指出来一点：“不要手动拉取项目的子模块”  要使用这层楼说明的方式递归地拉取项目，这样更省力，而且保证了所有子模块的版本是项目指定的版本。

**Reply by RCLiLong** [reply]
非常感谢指出这个问题，我将尽快更正，以免耽误他人。

**Reply by pugaizai** [reply]
@xfangfang 可以同时为x11和wayland编译依赖吗？

**Reply by xfangfang** [reply]
> @xfangfang 可以同时为x11和wayland编译依赖吗？

我也不是很清楚， wiliwili支持glfw或sdl，如果他们某一个支持的话wiliwili就能支持

**Reply by pugaizai** [reply]
GNOME Wayland桌面 Fedora 39 Silverblue
编译选项
```
      - '-DPLATFORM_DESKTOP=ON'
      - '-DUSE_SYSTEM_CURL=ON'
      - '-DINSTALL=ON'
      - '-DCMAKE_BUILD_TYPE=Release'
      - '-DGLFW_BUILD_WAYLAND=ON'
      - '-DGLFW_BUILD_X11=ON'
```
看起来没什么问题

**Reply by xfangfang** [reply]
@madoka773 非常感谢。
大体上看着很棒，有几处修改我有点小疑问：
1. 去掉了 LC_NUMERIC 的环境变量，我注意到mpv在flathub的脚本是保留这个变量的，对wiliwili来说下一个版本可以明确不需要这个环境变量（我在代码里强制设置了）但是对于1.2.2来说貌似还是应该有的？（因为我收到过一些wayland用户的反馈）
2. runtime 改到了 23.08，mpv 的runtime 还是 22.08， 是不是和mpv保持一致会更好一些？
3. finish-args 去除了配置目录设置（这个我也不是很确定需不需要，但是我注意到mpv写了，所以我当时也写了）
4. 另外一个无关的问题是，或许可以再发一条英文issue，帮助flathub审核人员确认得到了wiliwili开发者的授权。

太感谢您的贡献了～

**Reply by xfangfang** [reply]
@madoka773 我看了一下其他的pr，应该在 bot, build 后面加应用id的

**Comment by xfangfang** [comment]
树莓派 Raspberry Pi OS 请参考：https://github.com/xfangfang/wiliwili/issues/78#issuecomment-1416719288

使用 apt 的 debian 系操作系统，比如 ubuntu 也可以参考看看。

**Comment by xfangfang** [comment]
Steam Deck / Steam OS

```
# setup pacman
sudo steamos-readonly disable
sudo pacman-key --init
sudo pacman-key --populate archlinux
sudo pacman-key --populate holo

# install missing dependencies
sudo pacman -S cmake make gcc pkgconf xorgproto libxext mbedtls mpv libwebp zlib

# these dependencies are already installed, but you should reinstall them again
# refer to: https://www.reddit.com/r/SteamDeck/comments/t92ozw/for_compiling_c_code/
sudo pacman -S glibc linux-api-headers libx11 libxrandr libxinerama libxcursor libxi libxfixes libxrender dbus
```

**Comment by xfangfang** [comment]
NintendoSwitch / Switchroot Ubuntu

支持硬解 · 打包成AppImage

请参考：https://gist.github.com/xfangfang/c253580fb036c73549cbd237634c6d52
